### PR TITLE
[flang][cuda] Make default.nonTbpDefinedIoTable compiler generated

### PR DESCRIPTION
**Pull request by clementval**
`default.nonTbpDefinedIoTable` is a special global defined for IO that doesn't follow the mangling scheme and is then not handle correctly in the `CompilerGeneratedNames` pass. Update how it is generated with doGenerated so it can be handle without special handling. 

Also do not generate comdat in gpu module as the current code is not handling nested module correctly.

### CODE REVIEW

**Comment by jeanPerier** [comment]
Maybe "default.nonTbpDefinedIoTable" name could also be generated as a compiler generated name using fir::NameUniquer::doGenerated [here](https://github.com/llvm/llvm-project/blob/53d080c5b5dfbb46eb81d189736864f5b6196492/flang/lib/Lower/IO.cpp#L302) so that it later falls into the "Namekind::GENERATED" which is more accurate.

File: flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp [hunk 1]
```diff
@@ -48,8 +48,8 @@ void CompilerGeneratedNamesConversionPass::runOnOperation() {
     auto symName = op.getAttrOfType<mlir::StringAttr>(
         mlir::SymbolTable::getSymbolAttrName());
     auto deconstructedName = fir::NameUniquer::deconstruct(symName);
-    if (deconstructedName.first != fir::NameUniquer::NameKind::NOT_UNIQUED &&
-        !fir::NameUniquer::isExternalFacingUniquedName(deconstructedName)) {
+    if ((deconstructedName.first != fir::NameUniquer::NameKind::NOT_UNIQUED &&
+         !fir::NameUniquer::isExternalFacingUniquedName(deconstructedName))) {
```
Comment (clementval) [inline flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp:52]:
```suggestion
    if (deconstructedName.first != fir::NameUniquer::NameKind::NOT_UNIQUED &&
         !fir::NameUniquer::isExternalFacingUniquedName(deconstructedName)) {
```

File: flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp [hunk 2]
```diff
@@ -49,7 +49,7 @@ void CompilerGeneratedNamesConversionPass::runOnOperation() {
         mlir::SymbolTable::getSymbolAttrName());
     auto deconstructedName = fir::NameUniquer::deconstruct(symName);
     if (deconstructedName.first != fir::NameUniquer::NameKind::NOT_UNIQUED &&
-        !fir::NameUniquer::isExternalFacingUniquedName(deconstructedName)) {
+         !fir::NameUniquer::isExternalFacingUniquedName(deconstructedName)) {
```
Comment (clementval) [inline flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp:52]:
```suggestion
        !fir::NameUniquer::isExternalFacingUniquedName(deconstructedName)) {
```

**Comment by clementval** [comment]
> Maybe "default.nonTbpDefinedIoTable" name could also be generated as a compiler generated name using fir::NameUniquer::doGenerated [here](https://github.com/llvm/llvm-project/blob/53d080c5b5dfbb46eb81d189736864f5b6196492/flang/lib/Lower/IO.cpp#L302) so that it later falls into the "Namekind::GENERATED" which is more accurate. [edit: that way I believe not special handling would be needed here]

Yeah that was my other proposition. Just updated the patch so it's cleaner.

**Review by razvanlupusoru — APPROVED** [review]
LGTM!

**Pull request merged**